### PR TITLE
Fix `dask.array.fft` mismatch with Numpy's interface (add support for `norm` argument)

### LINKED ISSUE #3280: Support norm in dask.array.fft
The NumPy FFT functions optionally take a [`norm` argument]( https://docs.scipy.org/doc/numpy-1.14.0/reference/routines.fft.html#normalization ). Would be good if the Dask Array functions also optionally took that argument and only passed it on if it was not `None` (the default). This because it is easier to support other functions that don't take `norm` if we don't require that argument. Alternatively, this might be best addressed as arbitrary keyword argument as proposed in issue ( https://github.com/dask/dask/issues/3278 ).

### PR DESCRIPTION
**Pull request by joanrue**
- Closes  #3280 where `norm` argument was needed to match numpy.fft API. 
- Adds tests to validate that the different values for `norm` match NumPy's.
- Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Comment by joanrue** [comment]
Hi! I'm reaching out to see if there's anything more I can do to facilitate the review process.

**Comment by joanrue** [comment]
Hi ! The only failing test is not related to these changes. Can we go ahead with the PR?

**Comment by lucascolley** [comment]
this would be useful for SciPy! x-ref https://github.com/data-apis/array-api-compat/pull/139

File: dask/array/fft.py [hunk 1]
```diff
@@ -146,10 +146,8 @@ def fft_wrap(fft_func, kind=None, dtype=None):
     >>> parallel_ifft = dff.fft_wrap(np.fft.ifft)
     """
     if scipy is not None:
-        if fft_func is scipy.fftpack.rfft:
-            raise ValueError("SciPy's `rfft` doesn't match the NumPy API.")
-        elif fft_func is scipy.fftpack.irfft:
-            raise ValueError("SciPy's `irfft` doesn't match the NumPy API.")
+        if fft_func.__module__.startswith("scipy.fftpack"):
```
Comment (phofl) [inline dask/array/fft.py:149]: Why this change?

Comment (joanrue) [reply dask/array/fft.py:149]: Thanks for reviewing @phofl .

If we ignore the norm argument, only `rfft` and `irfft` were considered not to match NumPy's API. However, if we do consider the norm argument (as in this PR), then none of the functions from `scipy.fftpack` (including `fft`, `ifft`, `fft2`, etc.) match NumPy's API. 

This changes raises an error for all those functions from the depracated `scipy.fftpack` (including `rfft` and `irfft`).

Comment (phofl) [reply dask/array/fft.py:149]: got it, thanks. Can you include the function that is called in the error message to make it more descriptive?

Comment (phofl) [reply dask/array/fft.py:149]: > we ignore the norm argument, only rfft and irfft were considered not to match NumPy's API. However, if we do consider the norm argument (as in this PR), then none of the functions from scipy.fftpack (including fft, ifft, fft2, etc.) match NumPy's API.

Can we keep backwards compatibility somehow? Dask supports older version of scipy as well

Comment (phofl) [reply dask/array/fft.py:149]: 1.7.2, did it start raising deprecation warnings back then?

Comment (joanrue) [reply dask/array/fft.py:149]: Keeping backward compatibility would mean to replicate the behavior of `norm={“backward”, “ortho”, “forward”}` of `scipy.fft` functions as a patch for `scipy.fftpack` calls.

Comment (lucascolley) [reply dask/array/fft.py:149]: > did it start raising deprecation warnings back then?

It doesn't raise deprecation warnings - it is just marked as legacy in the docs.

**Comment by lucascolley** [comment]
@phofl @joanrue if we want to maintain support for `scipy.fftpack` without the `norm` argument, how about we:

- Create a new kwarg `allow_fftpack` for `fft_wrap`.
- Start raising a warning instead of an error here for `scipy.fftpack` input, noting that the NumPy API is not matched for the `norm` keyword.
  - Direct the user to use `scipy.fft` instead.
  - If the user still wants to use `scipy.fftpack`, make them pass `allow_fftpack` to silence the warning.
- Handle the failing `norm` with `scipy.fftpack` via `try ... except` or an explicit guard on the `fft_func.__module__` by default.
- Ignore `norm` when `allow_fftpack` is `True`

**Comment by phofl** [comment]
Yeah this sounds like a good path forward. I am fine with removing the usage of the legacy module at some point, but we should definitely deprecate this first and not just raise if scipy isn't raising warning either.

**Comment by joanrue** [comment]
@lucascolley @phofl thanks for your feedback!

**Review by lucascolley — APPROVED** [review]
thanks @joanrue, looks good to me! (I'm not familiar with the Dask codebase, so take that with a pinch of salt!)

File: dask/array/fft.py [hunk 2]
```diff
@@ -146,10 +146,18 @@ def fft_wrap(fft_func, kind=None, dtype=None):
...
+            if not allow_fftpack:
+                warnings.warn(
+                    f"Function {fft_func.__name__} from `scipy.fftpack` does not "
+                    "match NumPy's API and is considered legacy. Please use "
+                    "`scipy.fft` instead. To suppress this warning and allow usage"
+                    ", set `allow_fftpack=True`. Support for `scipy.fftpack` will "
+                    "be deprecated in future releases.",
+                    DeprecationWarning,
```
Comment (phofl) [inline dask/array/fft.py:157]: Feel free to make this a FutureWarning

Can you add a test that checks this warning?

**Review by phofl — COMMENTED** [review]
small comment

File: dask/array/fft.py (hunk 2 above)
Comment (joanrue) [reply dask/array/fft.py]: Good points, all done!

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx

**Comment by lucascolley** [comment]
@lithomas1 @asmeurer I think this should enable progress on https://github.com/data-apis/array-api-compat/pull/139 ?

**Comment by lithomas1** [comment]
Thanks, I'll take a look.
